usecases/config/runtime/manager.go

```diff
@@ -0,0 +1,189 @@
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package runtime
+
+import (
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	ErrEmptyConfig         = errors.New("empty runtime config")
+	ErrFailedToOpenConfig  = errors.New("failed to open runtime config")
+	ErrFailedToReadConfig  = errors.New("failed to read runtime config ")
+	ErrFailedToParseConfig = errors.New("failed to parse runtime config ")
+)
+
+// Parser takes care of unmarshaling a config struct
+// from given raw bytes(e.g: YAML, JSON, etc).
+type Parser[T any] func([]byte) (*T, error)
+
+// ConfigManager takes care of periodically loading the config from
+// given filepath for every interval period.
+type ConfigManager[T any] struct {
+	// path is file path of config to load and unmarshal from
+	path string
+	// interval is how often config manager trigger loading the config file.
+	interval time.Duration
+	// parse takes care of unmarshaling the config struct from a file
+	parse Parser[T]
+
+	// currentConfig is last successfully loaded config.
+	// ConfigManager keep using this config if there are any
+	// failures to load new configs.
+	currentConfig *T
+	currentHash   string
+	mu            sync.RWMutex // protects currentConfig
+
+	log             logrus.FieldLogger
+	lastLoadSuccess prometheus.Gauge
+	configHash      *prometheus.GaugeVec
+}
+
+func NewConfigManager[T any](
+	filepath string,
+	parser Parser[T],
+	interval time.Duration,
+	log logrus.FieldLogger,
+	r prometheus.Registerer,
+) (*ConfigManager[T], error) {
+	// catch empty filepath early
+	if len(strings.TrimSpace(filepath)) == 0 {
+		return nil, errors.New("filepath to load runtimeconfig is empty")
+	}
+
+	cm := &ConfigManager[T]{
+		path:     filepath,
+		parse:    parser,
+		interval: interval,
+		log:      log,
+		lastLoadSuccess: promauto.With(r).NewGauge(prometheus.GaugeOpts{
+			Name: "weaviate_runtime_config_last_load_success",
+			Help: "Whether the last loading attempt of runtime config was success",
+		}),
+		configHash: promauto.With(r).NewGaugeVec(prometheus.GaugeOpts{
+			Name: "weaviate_runtime_config_hash",
+			Help: "Hash value of the currently active runtime configuration",
+		}, []string{"sha256"}), // sha256 is type of checksum and hard-coded for now
+	}
+
+	// try to load it once to fail early if configs are invalid
+	if err := cm.loadConfig(); err != nil {
+		return nil, err
+	}
+
+	return cm, nil
+}
+
+// Run is a blocking call that starts the configmanager actor. Consumer probably want to
+// call it in different groutine. It also respects the passed in `ctx`.
+// Meaning, cancelling the passed `ctx` stops the actor.
+func (cm *ConfigManager[T]) Run(ctx context.Context) error {
+	return cm.loop(ctx)
+}
```

Comment:
why not launch the goroutine yourself in the constructor, and then provide the Close() method to shutdown the reading?

I think this pattern is less error-prone as people might accidentally forget to start the background sync

Reply:
I like to move concurrency to the client side as much as possible(to avoid hidden goroutines via API) . same as `http.Serve` blocking call